### ISSUE TITLE
Reliability issues

### ISSUE THREAD

**Issue by NathanKell** [KSP-RO/RealismOverhaul#2505]
Think we probably need to police reliability data. Issues I'm skeptical of / worried about:
- [ ] Are we double-dipping on failure chance? 80% cycle reliability and 80% ignition chance is 64% _actual_ success chance, e.g. (It's actually slightly lower due to the 5s of increased failure chance)
- [ ] If possible we should tweak the configger to take the increased failure chance in the first 5s into account. Right now it's a uniform (but inversely proportional to burn time) increase in failure chance.
- [ ] Some configs just don't make sense. The Altair starts at like 99.8% reliability but the Antares starts very low, e.g.
- [ ] The LR79 has a serious problem where S-3D is cheaper but has better specs, *and* better reliability than the config that unlocks one node later (LR79-NA-9).
- [ ] In general, IMO we should not simply have exactly the same values for cycle reliability and ignition chance. Pressure-fed hypergolics, e.g., should have rather higher ignition chance and lower cycle reliability, and the reverse for pump-fed engines I think.
- [ ] Another issue with ignitions: subsequent relights should probably have a higher success chance. That might need a TF change however.
- [ ] We might want to include some level of tech transfer between branches. E.g. XLR41 data -> XLR43 -> LR79/89/105, but also LR87 tree should probably inherit from XLR43 since it does for the ECM. Antares should inherit a bit from Altair and GCRC (and probably Sergeant too). Not much, like 5% rather than 50%, but some. With TF's branch support, it will work fine, e.g. LR79-NA-13 gets 50% of previous LR79 data _or_ 10% of LR87 data, _or_ 20% of XLR43 data, etc.

Anything else?

**Comment by leudaimon** [comment]
We could define a consistent way to deal with engines that have too few data for a good estimation of reliabilities, I have the impression this is all over the place

**Comment by jwvanderbeck** [comment]
> Are we double-dipping on failure chance? 80% cycle reliability and 80% ignition chance is 64% actual success chance, e.g. (It's actually slightly lower due to the 5s of increased failure chance)

You have a 20% chance to fail one check and a 20% chance to fail the other.  They aren't cumulative because one isn't dependent on the other. (one doesn't influence the other)

**Comment by StonesmileGit** [comment]
If the probabilities are completely independent, that means you can mutiply them, P(igniting) x P(burn for t seconds)

**Comment by StonesmileGit** [comment]
Two more things:

- [ ] Solids shouldn't get their ingition chance changed by dynamic pressure.
- [ ] The starting reliability of the Castor 1 is too high, it should get transfer from the first config.

**Comment by NathanKell** [comment]
> > Are we double-dipping on failure chance? 80% cycle reliability and 80% ignition chance is 64% actual success chance, e.g. (It's actually slightly lower due to the 5s of increased failure chance)
> 
> You have a 20% chance to fail one check and a 20% chance to fail the other. They aren't cumulative because one isn't dependent on the other. (one doesn't influence the other)

To put this another way: say an engine is classed as "98% success chance". If you naively set both cycle reliability and ignition reliability to 98%, the _actual_ success chance becomes (probability of ignition, 98%) x (probability of making it through the burn, 98%) = 96%. You've just doubled the failure chance. If in fact the engine is usually involved in a two-burn flight it's worse, 94.1%.

**Comment by marsh1832** [comment]
To add, they are cumulative in the sense where the ignition chance must pass before the burn chance can be checked. So the failure rate is cumulative. And as NK showed in an example of a second ignition, then that’s a third check in a row which must pass, so the mission failure rate is all three cumulative.

I love the idea of the failure rates being tuned to where the pump fed engines have a higher ignition failure but lower burn failure, and the inverse for pressure fed. Adds more character to the engine feed choice.

**Comment by NathanKell** [comment]
I have updated this issue to mention tech transfer and branches.

**Comment by Capkirk123** [comment]
#2539 addresses most of these issues.
For the reliability of pump fed vs pressure fed engines, that depends a lot on the engine itself. For example, Soviet pump-fed uppers almost never failed to ignite, but had their turbopumps fail during or shortly after their burn pretty frequently. The exception being the RD-58, which explodes on ignition to this day, but is usually fine once it starts. Even solids aren't immune, some had a propensity to explode, while other would just fail to ignite.

**Comment by ec429** [comment]
> If possible we should tweak the configger to take the increased failure chance in the first 5s into account.

FWIW TestLite already accounts for this internally: if an engine has a failure chance of (say) 10% across its 120-second rated burn time, then there will be a 1.5% chance of failure in the first 5s and an 8.5% chance in the next 120s, for 10% total over the 125s.

> you should get the full transfer only from the immediately prior config

Doesn't `techTransferGenerationPenalty` already cover that (to some extent)?  (As long as the configs are listed in the right order…)

**Comment by NathanKell** [comment]
No. Consider running an XLR43 until it's 10k data and then using an RS56.
You should not start at 2000 data for the RS56.

Sent by my thumbs, slowly.

**Comment by ec429** [comment]
Ah fair point, you're thinking of the max transfer.  But branches don't fix that either; they, like `techTransferGenerationPenalty`, only set the transfer coefficient.  The cap (`techTransferMax`) is the same no matter which config the transfer is coming from.

Maybe we need to add a `techTransferSourceMax`, that caps the du _before_ the branch modifier is applied, so that weaker links can't supply as much even if you have 10k on the old part.

(Either that or you're just referring to the fact that RO sets `techTransferGenerationPenalty` stupidly low, at 0.05.  Why isn't that higher?)

**Comment by NathanKell** [comment]
My point is that afaik generation penalty only applies when you fly
multiple generations. In the example I cite it wouldn't be used, the
transfer is direct from original to current part. How would the game know
those parts are many generations apart?

Sent by my thumbs, slowly.

**Comment by ec429** [comment]
Generation penalty is nothing to do with indirect transfers.  Generations are the comma-separated items within a `techTransfer` branch.  (In TL, indirect transfers don't happen at all.)

The breakage in your example only happens because the config is written backwards (#1968).  LR89-NA-7.1 has:

`techTransfer = XLR43-NA-1,A-6,A-6H,A-7:20&XLR43-NA-3,LR43-NA-3,LR89-NA-3,LR89-NA-5,LR89-NA-6:50`

which means it gets 50% transfer from XLR43-NA-3 but only 40% (`0.50 * (1 - (4 * 0.05))`) from LR89-NA-6.  (Under TestLite it's 30% instead (`0.50 - 4 * 0.05`) because I did things Differently for some reason.)

If it were written correctly as

`techTransfer = A-7,A-6H,A-6,XLR43-NA-1:20&LR89-NA-6,LR89-NA-5,LR89-NA-3,LR43-NA-3,XLR43-NA-3:50`

then the percentages would be the other way round and XLR43-NA-3 would be four generations away.

**Comment by NathanKell** [comment]
Ohh gotcha! Yeah that solves the problem

Sent by my thumbs, slowly.
